### PR TITLE
feat(widget): validate partner fee

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

1. Added controls in the widget configurator for partner fee percent and recipient address
2. Added validation on the app side for the fee percent and address
3. When there are some errors on the widget parameters then block the whole app and display the errors. @fairlighteth we need your power here 🙏 

image

image

  # To Test

1. Set a partner fee bellow 0%
- [ ] "Partner fee can not be less than 0%!" error should be displayed
2. Set a partner fee above 1%
- [ ] "Partner fee can not be more than 1%" error should be displayed
3. When the recipient is specified and not a valid address
- [ ] Then corresponding error should be displayed

# Questions

1. Should a recipient address be always specified along with the fee?

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/injectedWidget/consts.ts [hunk 1]
```diff
@@ -1 +1,3 @@
 export const COW_SWAP_WIDGET_EVENT_KEY = 'cowSwapWidget'
+
+export const PARTNER_FEE_MAX_BPS = 5000 // 50%
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/injectedWidget/consts.ts:3]: TODO: adjust the value

File: apps/cowswap-frontend/src/modules/injectedWidget/pure/WidgetParamsErrorsScreen/index.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  min-height: 300px;
+`
+
+export function WidgetParamsErrorsScreen({ errors }: { errors: WidgetParamsErrors }) {
+  const paramsWithErrors = Object.keys(errors)
+
+  if (paramsWithErrors.length === 0) return null
+
+  return (
+    <Container>
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/injectedWidget/pure/WidgetParamsErrorsScreen/index.tsx:25]: TODO: add beauty here

**Review by anxolin — COMMENTED** [review]
I see something strange in the values:

image

**Review by anxolin — COMMENTED** [review]
Should the error be expressed also in BPS instead of percentage? 
No more than 100 BPS

image

**Comment by elena-zh** [comment]
Hey @shoom3301 , I've got a crash when I filled in 0.001 value 
image

**Comment by elena-zh** [comment]
As for the Recipient validation, can you please make the address symbols case insensitive?  
image


Not a top priority, but when an error is displayed, it would be great to show it's actual value (if changed)
image

Question: is this OK to allow users to fill in 0x0000000000000000000000000000000000000000 address?
image

**Comment by shoom3301** [comment]
> I see something strange in the values:
> 
> image

@anxolin Sorry, 1% of 1000 should be 10.
1000 = 100%
   x     = 1%

 x = 1000 * 1 / 100
 x = 10

**Comment by shoom3301** [comment]
>Should the error be expressed also in BPS instead of percentage?
>No more than 100 BPS

@anxolin fixed it, thanks!

**Comment by shoom3301** [comment]
@elena-zh thank you for the feedback!
1. I think recipient always myst be set if partner fee is specified, so I added "Partner fee recipient must be set!" error
2. We don't support ENS here
3. Disabled a possibility to paste decimals in the BPS input
4. Recipient address validation must be case sensitive
5. Also fixed the error displaying (change it reactively)
6. 0x0000...0000 is a valid recipient address, I've even set it as default

**Comment by elena-zh** [comment]
Hey @shoom3301 , great, thank you!
The only one nitpick is that I'd not to show recipient error message once fee is set to 0
image

**Comment by shoom3301** [comment]
@elena-zh fixed

**Review by elena-zh — APPROVED** [review]
Great, thank you!

**Comment by anxolin** [comment]
Closes #3977 #3975

File: apps/widget-configurator/src/app/configurator/controls/PartnerFeeControl.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+export function PartnerFeeControl(props: PartnerFeeControlProps) {
+  const { feeBpsState, recipientState } = props
+  const [feeBps, setFeeBps] = feeBpsState
+  const [recipient, setRecipient] = recipientState
+
+  const onFeeBpsChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const value = Math.ceil(Number(e.target.value || 0))
+
+    setFeeBps(value)
+  }
```
Comment (alfetopito) [inline apps/widget-configurator/src/app/configurator/controls/PartnerFeeControl.tsx:18]: useCallback?

**Comment by anxolin** [comment]
> should Recipient field should be mandaroty when a user fills in the Partner fee field?

Yes! In the configurator I don't care too much, we could even potentially hide it as https://github.com/cowprotocol/cowswap/issues/3977 suggest. Although as it is now is ok IMO

@shoom3301 should we kill the `COMING SOON` messages? or we leave this for the launch?

**Review by anxolin — APPROVED** [review]

File: apps/widget-configurator/src/app/configurator/hooks/useWidgetParamsAndSettings.ts [hunk 4]
```diff
@@ -59,6 +61,14 @@ export function useWidgetParams(configuratorState: ConfiguratorState, isDappMode
       },
 
       hideConnectButton: isDappMode,
+
+      partnerFee:
+        partnerFeeBps && partnerFeeBps > 0
```
Comment (alfetopito) [inline apps/widget-configurator/src/app/configurator/hooks/useWidgetParamsAndSettings.ts:66]: Is the first part needed?

`undefined > 0 === false`

**Review by alfetopito — APPROVED** [review]
Minor nitpicks

**Comment by shoom3301** [comment]
> > should Recipient field should be mandaroty when a user fills in the Partner fee field?
> 
> Yes! In the configurator I don't care too much, we could even potentially hide it as #3977 suggest. Although as it is now is ok IMO
> 
> @shoom3301 should we kill the `COMING SOON` messages? or we leave this for the launch?

@anxolin good point, I will remove them in another PR

File: apps/cowswap-frontend/src/modules/injectedWidget/utils/validatePartnerFee.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+import { PARTNER_FEE_MAX_BPS } from '../consts'
+
+export function validatePartnerFee(input: PartnerFee | undefined): string[] | undefined {
+  if (!input) return undefined
+
+  const { bps, recipient } = input
+
+  const feeTooHighError =
+    bps > PARTNER_FEE_MAX_BPS ? `Partner fee can not be more than ${PARTNER_FEE_MAX_BPS / 100}%!` : undefined
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/injectedWidget/utils/validatePartnerFee.ts]: Fixed

File: apps/widget-configurator/src/app/configurator/controls/PartnerFeeControl.tsx (hunk 3 above)
Comment (shoom3301) [reply apps/widget-configurator/src/app/configurator/controls/PartnerFeeControl.tsx:18]: Fixed

File: apps/widget-configurator/src/app/configurator/hooks/useWidgetParamsAndSettings.ts (hunk 4 above)
Comment (shoom3301) [reply apps/widget-configurator/src/app/configurator/hooks/useWidgetParamsAndSettings.ts:66]: Fixed

File: apps/cowswap-frontend/src/modules/injectedWidget/utils/validateWidgetParams.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+export function validateWidgetParams(params: CowSwapWidgetParams): WidgetParamsErrors {
+  const keys = Object.keys(params) as Keys[]
+
+  return keys.reduce<WidgetParamsErrors>((acc, key) => {
+    const validation = VALIDATIONS[key]
+
+    if (!validation) return acc
+
+    const errors = validation(params[key] as never)
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/injectedWidget/utils/validateWidgetParams.ts:21]: To be honest, I couldn't beat typescript here :( 
I will be happy to improve it, but have no Idea how to do that.

image

**Pull request merged**